clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2247,6 +2251,22 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::get_input_last_timestamp(
     return res;
 }
 
+template <typename RecordType, typename ReaderType>
+uint64_t
+scheduler_impl_tmpl_t<RecordType, ReaderType>::get_next_trace_pc(output_ordinal_t output)
+{
+    if (output < 0 || output >= static_cast<output_ordinal_t>(outputs_.size()))
+        return 0;
+    int index = outputs_[output].cur_input;
+    if (index < 0)
+        return 0;
+    if (inputs_[index].cur_from_queue &&
+        inputs_[index].cur_queue_record.next_trace_pc != 0) {
```

Comment:
Why do we need this 2nd check, !=0? Shouldn't we always use the next PC for this entry?

If a next PC of 0 is an error, it should be an error to !=0 and we should raise an error instead of returning the next-next PC (asking the input below).

If a next PC of 0 is not an error, we should return 0.

Reply:
`cur_queue_record.next_trace_pc = 0` means that a next_trace_pc value snapshot was not taken for this record, and the scheduler should consult the input stream directly. 

E.g., insert_switch_tid_pid adds stuff to the queue but doesn't add the next trace pc snapshot, because the input stream can directly provide it.

The only case I found where the next_trace_pc needs to be remembered for the records read ahead at the beginning.

Reply:
It seems better to rely on the reader APIs as much as we can, and avoid bypassing it when we can.